byzcoin/service.go

```diff
@@ -427,14 +423,76 @@ func (s *Service) AddTransaction(req *AddTxRequest) (*AddTxResponse, error) {
 		log.Lvlf2("Instruction[%d]: %s on instance ID %s", i, instr.Action(), instr.InstanceID.String())
 	}
 
+	//Either send the transaction to the leader, or, if this node is the leader, directly send it to ctxChan.
+	//For every new tx create a new protocol, like in skipchain
+
+	leader, err := s.getLeader(req.SkipchainID)
+	if err != nil {
+		return nil, xerrors.Errorf("Error getting the leader", err)
+	}
+
+	ctxHash := req.Transaction.Instructions.Hash()
+
+	if s.ServerIdentity().Equal(leader) {
+		s.txPipeline.ctxChan <- req.Transaction
+		if header.Version < req.Version {
+			s.txPipeline.needUpgrade <- req.Version
+		}
+	} else {
+		latest, err := s.db().GetLatestByID(req.SkipchainID)
+		if err != nil {
+			log.Errorf("Error while searching for %x", req.SkipchainID[:])
+			log.Error("DB is in bad state and cannot find skipchain anymore."+
+				" This function should never be called on a skipchain that does not exist.", err)
+			return nil, xerrors.Errorf("reading latest: %v", err)
+		}
+
+		//create new roster with self and leader
+		list := make([]*network.ServerIdentity, 0)
+		list = append(list, []*network.ServerIdentity{s.ServerIdentity(), leader}...)
+		newRost := onet.NewRoster(list)
+		tree := newRost.GenerateNaryTree(len(newRost.List))
+
+		proto, err := s.CreateProtocol(rollupTxProtocol, tree)
+		if err != nil {
+			log.Error(s.ServerIdentity(), "Protocol creation failed with error."+
+				" This panic indicates that there is most likely a programmer error,"+
+				" e.g., the protocol does not exist."+
+				" Hence, we cannot recover from this failure without putting"+
+				" the server in a strange state, so we panic.", err)
+			return nil, xerrors.Errorf("creating protocol: %v", err)
+		}
+
+		root := proto.(*RollupTxProtocol)
+		root.SkipchainID = req.SkipchainID
+		root.LatestID = latest.Hash
+		root.NewTx = req
+		err = proto.Start()
+		if err != nil {
+			return nil, xerrors.Errorf("Error starting the protocol", err)
+		}
+		if err := <-root.DoneChan; err != nil {
```

Comment:
Perhaps it would be safer to `select` with a timeout, to avoid waiting forever

Reply:
This was changed